### DISCUSSION TITLE
Why can't I get the latest value of useState in subscribe?

### DISCUSSION THREAD

**Discussion by ru88s** [supabase/supabase#3493] [Questions] [answered]
Why can't I get the latest value of useState in subscribe?
I want to use the latest value when data is added.

```js
import { useEffect, useState } from "react";
import { useLocalStorage } from "react-use";
import { supabase } from "src/lib/initSupabase";

const Demo = () => {
  const [editMode, setEditMode] = useState(false);
  const [commentTracking, setCommentTracking] = useLocalStorage("commentTracking", false);

  useEffect(() => {
    supabase
      .from("comments")
      .on("INSERT", (payload) => {
        console.log(editModeParent);
        console.log(commentTracking);
      })
      .subscribe();
  }, []);

  return (
    <div>
      <button
        onClick={() => {
          return setEditMode(!editMode);
        }}
      >
        Button1
      </button>
      <button
        onClick={() => {
          return setCommentTracking(!commentTracking);
        }}
      >
        Button2
      </button>
    </div>
  );
};
```

**Comment by VictorPeralta** [comment]
This is an issue with stale closure in `useEffect`. As your code stands, the useEffect is only executed once after initial mount, and the values of your state will remain as it was when the effect ran.  To fix this, add `editMode` to your dependency array

[Here's a blog post that explains further](https://dmitripavlutin.com/react-hooks-stale-closures/)

**Comment by ru88s** [comment, marked as answer]
@VictorPeralta Thanks! 
When I add `editMode` to the array in this code and insert it, I get console.log output over and over again.
It is not possible to get the exact value.

```js
import { useEffect, useState } from "react";
import { useLocalStorage } from "react-use";
import { supabase } from "src/lib/initSupabase";

const Demo = () => {
  const [editMode, setEditMode] = useState(false);
  const [commentTracking, setCommentTracking] = useLocalStorage("commentTracking", false);

  useEffect(() => {
    supabase
      .from("comments")
      .on("INSERT", (payload) => {
        console.log("commentTracking: " + commentTracking);
      })
      .subscribe();
  }, [editMode, commentTracking]);

  return (
    <div>
      <button
        onClick={() => {
          return setEditMode(!editMode);
        }}
      >
        Button1
      </button>
      <button
        onClick={() => {
          return setCommentTracking(!commentTracking);
        }}
      >
        Button2
      </button>
    </div>
  );
};
```

![スクリーンショット 2021-10-13 15 57 02](https://user-images.githubusercontent.com/42270511/137082752-aa0d690d-ada6-4c14-a67c-28e6beb2e315.png)

**Reply by GaryAustin1** [reply]
With this new code it looks like you will get a new subscription every time the effect runs (buttons clicked?).  So you will start seeing multiple console.logs

**Reply by ru88s** [reply]
@GaryAustin1 Sorry, this is the more correct code.
When insert occurs, console.log is output again and again.
I want it to happen only once.

```js
import { useEffect, useState } from "react";
import { supabase } from "src/lib/initSupabase";

const Demo = () => {
  const [editMode, setEditMode] = useState(false);

  useEffect(() => {
    supabase
      .from("comments")
      .on("INSERT", (payload) => {
        console.log("editMode: " + editMode);
      })
      .subscribe();
  }, [editMode]);

  const dbInsert = async () => {
    setEditMode(!editMode);

    const { data, error } = await supabase.from("comments").insert([
      {
        body: "test",
      },
    ]);
  };

  return (
    <div>
      <button
        onClick={() => {
          return dbInsert();
        }}
      >
        Button
      </button>
    </div>
  );
};

```

**Reply by GaryAustin1** [reply]
I don't see an infinite loop, but every time you click the button to do an insert, the useffect is going to run (because you toggle editMode) and each time you will get an extra subscription to the same insert.  So code starts up you get a subscriber with no response, then you do first insert and get a 2nd subscriber, plus the first both responding and doing console.log.  Do another button click and now you have 3 subscribers.

**Reply by ru88s** [reply]
@GaryAustin1 Thanks. Does this mean that it is impossible to get the correct `eidtMode` value only once?

**Reply by GaryAustin1** [reply]
I really don't understand what you are doing with editmode as it just toggles between true and false.  If you just have the useEffect without any parameters in the [] it will run once and set up your subscription.  Then if you console.log out payload you will see data each time there is an insert and can do what you want with it.

**Reply by ru88s** [reply]
editmode has the same meaning as textarea focus.
We want to determine if text is being typed and auto-scroll to the added comment if it is not being typed.
If you run useEffect without any parameters in [], you will only get the old value of editMode.

**Reply by GaryAustin1** [reply]
At this point you are really dealing with more React stuff than Supabase.  I'm competent in React, but not an expert.  BUT you can't keep calling subscribe over and over as it just generates extra subscribes and will give you problems, So you need to rethink somehow splitting up the subscribe useEffect and what ever state you are trying to maintain in editmode.

**Reply by ru88s** [reply]
Sorry, I've improved the code.
I would like to determine the editMode and do scrolling.

```js
import { useEffect, useState } from "react";
import { supabase } from "src/lib/initSupabase";
import TextareaAutosize from "react-textarea-autosize";
import { scroller } from "react-scroll";

const Demo = () => {
  const [editMode, setEditMode] = useState(false);

  useEffect(() => {
    supabase
      .from("comments")
      .on("INSERT", (payload) => {
        console.log("editMode: " + editMode);

        if (!editMode) {
          scroller.scrollTo(String(payload.new.id), {});
        }
      })
      .subscribe();
  }, [editMode]);

  const dbInsert = async () => {
    const { data, error } = await supabase.from("comments").insert([
      {
        body: "test",
      },
    ]);
  };

  return (
    <div>
      <TextareaAutosize
        onFocus={() => {
          setEditMode(true);
        }}
        onBlur={() => {
          setEditMode(false);
        }}
      />
      <button
        onClick={() => {
          return dbInsert();
        }}
      >
        Button
      </button>
    </div>
  );
};
```

**Reply by GaryAustin1** [reply]
One more comment.  You might get further if you call a function from the subscribe payload routine and do whatever checking on editmode there versus editmode being part of useEffect as that causes it run again everytime editmode changes generating more subscriptions.

**Reply by VictorPeralta** [reply]
You're creating a new subscription every time the effect runs, but never unsubscribing. Do this in your useEffect's return statement

https://supabase.io/docs/reference/javascript/removesubscription

**Reply by ru88s** [reply]
@GaryAustin1 Thank you for your help many times.
I tried to do this but could not get the changed value.

@VictorPeralta Thanks. I added removeSubscription.
Would this be the correct code for performance?

```js
import { useEffect, useState } from "react";
import { supabase } from "src/lib/initSupabase";
import TextareaAutosize from "react-textarea-autosize";
import { scroller } from "react-scroll";

const Demo = () => {
  const [editMode, setEditMode] = useState(false);

  useEffect(() => {
    const mySubscription = supabase
      .from("comments")
      .on("INSERT", (payload) => {
        editModeCheck(payload);
      })
      .subscribe();

    return () => {
      if (mySubscription) supabase.removeSubscription(mySubscription);
    };
  }, []);

  const editModeCheck = (payload: any) => {
    console.log("editMode: " + editMode);
    if (!editMode) {
      scroller.scrollTo(String(payload.new.id), {});
    }
  };

  const dbInsert = async () => {
    const { data, error } = await supabase.from("comments").insert([
      {
        body: "test",
      },
    ]);
  };

  return (
    <div>
      <TextareaAutosize
        onFocus={() => {
          setEditMode(true);
        }}
        onBlur={() => {
          setEditMode(false);
        }}
      />
      <button
        onClick={() => {
          return dbInsert();
        }}
      >
        Button
      </button>
    </div>
  );
};
```

**Reply by VictorPeralta** [reply]
Your value won't change this way because of the same closure issue as you had before. To fix this, move the function declaration into the effect, or wrap the function in useCallback, with editMode in the useCallback dependency array and add it to the useEffect dependency array.

**Reply by VictorPeralta** [reply]
Also editMode should be included in the useEffect dependency array if you declare the function inside the effect

**Reply by ru88s** [reply]
@VictorPeralta Thanks.
If you do this, the subscription will not work because it will be removedSubscription at a bad time.
I'm looking into useCallback, but I don't know how to use it properly. I'm sorry.

```js
useEffect(() => {
    const mySubscription = supabase
      .from("comments")
      .on("INSERT", (payload) => {
        editModeCheck(payload);
      })
      .subscribe();

    return () => {
      if (mySubscription) supabase.removeSubscription(mySubscription);
    };
  }, [editMode]);
```

**Reply by VictorPeralta** [reply]
Now that I see what you're trying to do, it might be easier to accomplish it with useRef instead of useState
